[13805] Use Case - Record Replay with Rosbag (#326)
eProsima/Fast-DDS-docs: docs/fastdds/use_cases/rosbag_capture/rosbag_capture.rst

.. include:: ../../../03-exports/aliases.include
.. include:: ../../../03-exports/aliases-api.include
.. include:: ../../../03-exports/roles.include

.. _rosbag_capture:

Rosbag2 Record and Replay
==========================

> Rosbag2 is a ROS application that can be used to capture DDS messages and store them on an sqlite database. This allows 

Reviewer: In general, please break line after each sentence period

> Rosbag2 is a ROS application that can be used to capture DDS messages and store them on an sqlite database. This allows 

Reviewer: ```suggestion
Rosbag2 is a ROS 2 application that can be used to capture DDS messages and store them on an sqlite database. This allows 
```

inspecting and replaying said messages at a later time. 

> Rosbag2 interactions with a non-ROS2 Fast DDS application

Reviewer: ```suggestion
Rosbag2 interactions with a native Fast DDS application
```

---------------------------------------------------------

Using rosbag2 to capture traffic between ROS talkers and listeners is straightforward. Using it to record and replay
messages sent by Fast DDS participants outside ROS2 ecosystem requires some modifications. 

Prerequisites
^^^^^^^^^^^^^

> Being a ROS2 application, to install rosbag requires to have the ROS2 repository for the ROS2 distribution of choice

Reviewer: In general
```suggestion
Being a ROS 2 application, to install rosbag requires to have the ROS 2 repository for the ROS 2 distribution of choice
```

set up in your sources file. Instructions on how to do so can be found on the `ROS2 documentation
<https://docs.ros.org/en/galactic/Installation/Ubuntu-Install-Binary.html#add-the-ros-2-apt-repository>`_

Rosbag2 can then be installed following the instructions on their `Github repository
<https://github.com/ros2/rosbag2>`_

IDL files
^^^^^^^^^
IDL files used to generate types for the application need to have the types nested inside both the module name and an idl 
module, in that order, for instance:

> .. code-block:: bash

Reviewer: Please format IDL
```suggestion
.. code-block:: idl
```

   module helloworld_type{

>      module idl {

Reviewer: I'd clearly state that the IDL module must be named `idl`
```suggestion
     module idl {
```

"new_typesupport::idl::HelloWorld". It's easier however to use the accessor provided to this value during
the create_topic call. Something similar to:

-- code-block:: bash

   topic = participant->create_topic("rt/HelloWorldTopic", type->getName(), TOPIC_QOS_DEFAULT);

Rosbag Record and Play
^^^^^^^^^^^^^^^^^^^^^^
Using the overlay created earlier one can make ROS applications aware of our custom type. Sourcing the type
support package and then launching rosbag2 will allow our messages to be captured.

.. code-block:: bash

   export RMW_IMPLEMENTATION=rmw_fastrtps_cpp
   source new_typesupport/install/setup.bash
   ros2 bag record /TopicNameWithoutRT

Launch your publisher application and you will see a discovery on the rosbag log:

.. code-block:: bash

   [INFO] [1644320308.422292205] [rosbag2_recorder]: All requested topics are subscribed. Stopping discovery...

Rosbag will proceed to create a folder with an SQLite database inside. The path to this database file can be
used to replay the recorded messages. Open a subscriber application for the same topic and run:

.. code-block:: bash

   export RMW_IMPLEMENTATION=rmw_fastrtps_cpp

>    source new_typesupport/install/setup.bash

Author: Not necessarily true. The user might want to replay the messages from a different terminal or at a later date.
